Add address-based index (attempt 4?)

marcinja opened the pull request:
Adds index to transactions by scriptPubKey. Based off of #2802. Stores hashes of scripts (hashed using Murmurhash3, with a hash seed that is stored in the index database), along with the `COutPoint` for the output which was spent/created, and the `CDiskTxPos` for the transaction in which this happened.

ryanofsky commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,225 @@
+                    // be ahead of txindex. We let txindex sync first so that addrindex can continue
+                    // after it.
+                    while (!g_txindex->IsInSyncWithMainChain()) {
+                        MilliSleep(1000); //TODO: find a less arbitrary sleep time.
```
In commit "Introduce address index" (3c7cc3c705b08828dc8dc919e53343df78568ebd)

This sleep + BlockUntilSyncedToCurrentChain call is pretty hacky but probably ok in principle. I think @jimpo could easily tell you how to improve it. I would move this code out of the `AddrIndex` class and into a `g_txindex->WaitForSync()` or similar method, and probably replace the sleep with a condition variable wait.

Or maybe this code will be unnecessary with #14035 undo data?

practicalswift replied:
```
2018-09-22 21:20:09 cpplint(pr=14053): src/index/addrindex.cpp:124:  Should have a space between // and comment  [whitespace/comments] [4]
```

ryanofsky commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,225 @@
+    while (iter->Valid()) {
+        std::pair<std::pair<char, uint64_t>, CDiskTxPos> key;
+        uint64_t value;
+        if (!iter->GetKey(key) || !iter->GetValue(value) || key.first != key_prefix) break;
```
In commit "Introduce address index" (3c7cc3c705b08828dc8dc919e53343df78568ebd)

Should probably reorder this condition. It seems pointless to retrieve the value when the key is wrong, and maybe dangerous if the value is a different format and deserializing could throw an exception.

ryanofsky commented on src/rpc/rawtransaction.cpp:
```diff
@@ -203,6 +204,123 @@ static UniValue getrawtransaction(const JSONRPCRequest& request)
     return result;
 }
 
+static UniValue searchrawtransactions(const JSONRPCRequest& request) {
```
In commit "Add searchrawtransactions RPC" (226eeea9736127269058fbbf7816c7100f90974d)

Should add a basic python functional test for the new RPC method.

ryanofsky reviewed (commented):
Reviewed most of the code, but just skimmed tests. It looks to me like this PR could be merged basically in its current form, so I'm curious if you're intending to make the improvements cited in the PR description above here or in a separate PRs.

I left a few minor suggestions about the code, which you should feel free to ignore. The only changes I would definitely like to see here are:

- adding some python code in `test/functional/` to call the new rpc method
- adding a blurb in `doc/release notes.md` to describe the feature and maybe mention some use-cases

luke-jr commented on src/rpc/rawtransaction.cpp:
```diff
@@ -203,6 +204,129 @@ static UniValue getrawtransaction(const JSONRPCRequest& request)
+            "\nArguments:\n"
+            "1. \"address\"    (string, required) The address to search for\n"
+            "2. \"verbose\"    (bool, optional, default = false) If set to false, only returns data for hex-encoded `txid`s. \n"
+            "3. \"skip\"       (numeric, optional, default = 0) If set, the result skips this number of initial values. \n"
```
`skip` and `count` probably make sense on an options object instead.

luke-jr commented on src/rpc/rawtransaction.cpp:
```diff
@@ -203,6 +204,129 @@ static UniValue getrawtransaction(const JSONRPCRequest& request)
+            "1. \"address\"    (string, required) The address to search for\n"
+            "2. \"verbose\"    (bool, optional, default = false) If set to false, only returns data for hex-encoded `txid`s. \n"
+            "3. \"skip\"       (numeric, optional, default = 0) If set, the result skips this number of initial values. \n"
+            "3. \"count\"      (numeric, optional, default = 100) If set, the result will only contain this amount of values. \n"
```
(this would be number 4, but n/a once on an options object)

luke-jr commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,225 @@
+            positions.emplace_back(GetAddrID(tx_out.scriptPubKey), pos);
+        }
+
+        if (g_txindex && !tx->IsCoinBase()) {
```
This basically shouldn't be here. The address used to receive a coin has zero relationship with the transaction spending that coin.

If someone really cares about getting it anyway, there should be a separate input txid/index -> spend txid index.

luke-jr commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,225 @@
+                    g_txindex->BlockUntilSyncedToCurrentChain();
+
+                    // If we still can't find the tx then a re-org may have happened.
+                    if (!g_txindex->FindTx(tx_in.prevout.hash, block_hash, tx)) return false;
```
This looks like a very fragile assumption.

luke-jr commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,225 @@
+
+                    // It's also possible we can't find the tx in txindex because it fell behind in
+                    // the ValidationInterface queue. In this case we also let it finish before continuing.
+                    g_txindex->BlockUntilSyncedToCurrentChain();
```
With this, why is the loop waiting for IsInSyncWithMainChain needed?

marcinja replied:
`BlockUntilSyncedToCurrentChain` returns false immediately if `m_synced` is false.

The loop waiting for `IsInSyncWithMainChain` waits for txindex to finish in `ThreadSync`. Once txindex finishes in `ThreadSync` it will be updated by ValiditionInterface callbacks from that point on.

This line (since txindex must have `m_synced` =  true now) will actually block now until txindex syncs with the ValidationInterface queue.

marcinja commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,225 @@
+    return WriteBatch(batch);
+}
+
+void AddrIndex::BlockDisconnected(const std::shared_ptr<const CBlock> &block) {
```
This needs an `if (!m_synced) return;`

jimpo commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,225 @@
+
+bool AddrIndex::DB::ReadAddrIndex(const uint64_t addr_id,
+                                  std::vector<std::pair<std::pair<char, uint64_t>, CDiskTxPos>> &keys_found,
+                                  const bool filter_by_value,
```
Seems that `filter_by_value` functionality isn't used? Also, I'd rename to something like `filter_by_block` because it's not clear what value means and is tightly coupled to the database layout chosen.

jimpo reviewed (changes requested):
Nice work! I'm glad someone's working on this. Concept ACK.

1. The AddrIndex should return information about the outpoints and differentiate between outputs and spends, not just return the raw transactions. In fact, the AddrIndex could *just* return outpoints, then the client code could use the TxIndex could to fetch the tx bodies. It'd involve a separate lookup though.
2. I don't think it's necessary to delete keys from the database when a block is disconnected. There's no harm in leaving it. The higher level methods to search the index can then filter for results that are on the main chain if that's what the client wants. It'd have to do this anyway to avoid races with reorgs and such.
3. I'm worried about collisions on address IDs because they are only 64 bits. I can think of three options, 1) use a 32 byte cryptographic hash, 2) use a 20 byte cryptographic hash of the script plus some randomly generated, database-global salt, 3) use a 32- or 64-bit non-cryptographic hash (might as well use Murmur3 or SipHash, not truncated SHA256), then store the full script as the database value to double check against. Option 3 feels best to me.
4. What's the purpose of having the first byte of the block hash as the value? It doesn't seem robust nor particularly useful.

marcinja commented:
Thanks for all the reviews. 

To answer some of @jimpo's questions:
2 & 4. I included part of the block hash so that in BlockDisconnected we ae sure to remove the entries in the index from this block only (that's where  `filter_by_value` is used). The reason I chose to remove entries from the database is to prevent reading into a block file using an old `CDiskTxPos`  that may no longer be a valid position. Otherwise in `FindTxsByScript` you could run into errors. You're right that this problem would be better handled by higher level methods.

I think that returning just the outpoint is a better idea than the current choice so I'll switch to that and try to incorporate all the other feedback here.

practicalswift commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,225 @@
+    return m_db->WriteToIndex(positions, block.GetHash());
+}
+
+bool AddrIndex::DB::WriteToIndex(const std::vector<std::pair<uint64_t, CDiskTxPos>>& positions, const uint256 block_hash)
```
`block_hash` should be passed by const reference?

practicalswift commented on src/test/test_bitcoin.cpp:
```diff
@@ -173,6 +175,76 @@ TestChain100Setup::CreateAndProcessBlock(const std::vector<CMutableTransaction>&
+// Based off of BuildChain in validation_block_tests.cpp
+
+// Build a chain of blocks that contains all of the transactions in txns.
+void TestChain100Setup::BuildChain(const uint256 prev_hash, const uint32_t prev_time, int height, std::vector<CMutableTransaction> &txns, const CScript& scriptPubKey, std::vector<std::shared_ptr<const CBlock>>& blocks) {
```
`prev_hash` should be passed by const reference?

practicalswift commented on src/test/test_bitcoin.h:
```diff
@@ -89,6 +89,15 @@ struct TestChain100Setup : public TestingSetup {
     CBlock CreateAndProcessBlock(const std::vector<CMutableTransaction>& txns,
                                  const CScript& scriptPubKey);
 
+    void BuildChain(const uint256 prev_hash,
```
`prev_hash` should be passed by const reference?

practicalswift commented on src/test/test_bitcoin.cpp:
```diff
@@ -173,6 +175,76 @@ TestChain100Setup::CreateAndProcessBlock(const std::vector<CMutableTransaction>&
+    BuildChain(blocks.back()->GetHash(), blocks.back()->nTime, height - 1, txns, scriptPubKey, blocks);
+}
+
+void TestChain100Setup::CreateSpendingTxs(int coinbase_spent_offset, std::vector<CScript>& script_pub_keys, std::vector<CMutableTransaction> &spends, CScript coinbase_script_pub_key) {
```
`coinbase_script_pub_key` should be passed by const reference?

practicalswift commented on src/rpc/rawtransaction.cpp:
```diff
@@ -203,6 +204,129 @@ static UniValue getrawtransaction(const JSONRPCRequest& request)
+    }
+
+    std::vector<std::pair<uint256, CTransactionRef>>::const_iterator it = result.begin();
+    while (it != result.end() && skip--) it++; // Skip first set of results as needed.
```
Nit: `++it`

practicalswift commented on src/rpc/rawtransaction.cpp:
```diff
@@ -203,6 +204,129 @@ static UniValue getrawtransaction(const JSONRPCRequest& request)
+            tx_val.pushKV("blockhash", tuple.first.GetHex());
+        }
+        ret.push_back(tx_val);
+        it++;
```
Nit: `++it`

practicalswift commented on src/test/test_bitcoin.cpp:
```diff
@@ -173,6 +175,76 @@ TestChain100Setup::CreateAndProcessBlock(const std::vector<CMutableTransaction>&
+    // If this is the last block, add all remaining transactions.
+    // Otherwise add with some randomness.
+    for (auto it = txns.begin(); it != txns.end();) {
+        bool add_tx = (height == 1) || (GetRandInt(height) < txns.size());
```
```
test/test_bitcoin.cpp:194:60: warning: comparison of integers of different signs: 'int' and 'std::vector::size_type' (aka 'unsigned long') [-Wsign-compare]
        bool add_tx = (height == 1) || (GetRandInt(height) < txns.size());
                                        ~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~
```

practicalswift commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,225 @@
+
+    // Index addresses of spent outputs if txindex is enabled,
+    for (const auto& tx : block.vtx) {
+        for (const auto tx_out : tx->vout){
```
```
index/addrindex.cpp:110:25: warning: loop variable 'tx_out' of type 'const CTxOut' creates a copy from type 'const CTxOut' [-Wrange-loop-analysis]
        for (const auto tx_out : tx->vout){
                        ^
index/addrindex.cpp:110:14: note: use reference type 'const CTxOut &' to prevent copying
        for (const auto tx_out : tx->vout){
             ^~~~~~~~~~~~~~~~~~~
```

practicalswift commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,225 @@
+        }
+
+        if (g_txindex && !tx->IsCoinBase()) {
+            for (const auto tx_in : tx->vin) {
```
Same here - see above.

practicalswift commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,225 @@
+
+        // Collect all addr_ids from txs in this block.
+        for (const auto& tx : block->vtx) {
+            for (const auto tx_out : tx->vout){
```
Same here - see above.

practicalswift commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,225 @@
+            }
+
+            if (!tx->IsCoinBase()) {
+                for (const auto tx_in : tx->vin){
```
Same here - see above.

practicalswift commented on src/test/addrindex_tests.cpp:
```diff
@@ -0,0 +1,418 @@
+    CScript coinbase_script_pub_key = CScript() <<  ToByteVector(coinbaseKey.GetPubKey()) << OP_CHECKSIG;
+    for (int i = 0; i < 20; i++) {
+        std::vector<CMutableTransaction> no_txns;
+        const CBlock& block = CreateAndProcessBlock(no_txns, coinbase_script_pub_key);
```
```
test/addrindex_tests.cpp:112:23: warning: unused variable 'block' [-Wunused-variable]
        const CBlock& block = CreateAndProcessBlock(no_txns, coinbase_script_pub_key);
                      ^
```

Sjors reviewed (commented):
Concept ACK. Suggest adding `-addrindex` and `searchrawtransactions <address>` to the PR description for easier review.

I would do "integrating output descriptors" in a different PR.

Lightly tested on macOS. It doesn't find the coinbase transaction to `mtnNvZY7iQCfzJCVon13tfJSVYDn9iiUWD` in [block 120](https://testnet.blockchain.info/en/block/0000000058dbb5e3fce9be42ae79beb564be666866c6b85c9791863219ec36db), `mg4bQva8w7Cjs8KaKHeEbyXUaCzoQAcXyH`  in [block 10001](https://testnet.blockchain.info/nl/block/0000000000629d100db387f37d0f37c51118f250fb0946310a8c37316cbc4028) and several others.

practicalswift commented on src/index/addrindex.h:
```diff
@@ -0,0 +1,56 @@
+    explicit AddrIndex(size_t n_cache_size, bool f_memory = false, bool f_wipe = false);
+
+    // Destructor is declared because this class contains a unique_ptr to an incomplete type.
+    virtual ~AddrIndex() override;
```
```
2018-09-22 21:20:09 cpplint(pr=14053): src/index/addrindex.h:44:  "virtual" is redundant since function is already declared as "override"  [readability/inheritance] [4]
```

karel-3d commented:
Comments to the concept

Some time ago I have rebased bitpay's bitcore patches (from their insight block explorer) to bitcoin core here (without much work on my own)

https://github.com/bitcoin/bitcoin/pull/10370

The consensus back then seemed to be that more indexes in the core is not a good thing. However I am no sure how this PR differs.

Sjors commented:
We talked about (address) indexes during the previous [IRC meeting](http://www.erisian.com.au/meetbot/bitcoin-core-dev/2018/bitcoin-core-dev.2018-12-13-19.00.html).

practicalswift commented on src/test/addrindex_tests.cpp:
```diff
@@ -0,0 +1,154 @@
+
+    // Now check that all the addresses we sent to are present in the index.
+    for (int i = 0; i < 10; i++) {
+        std::vector<std::pair<COutPoint, std::pair<CTransactionRef, uint256>>> spends;
```
`spends` shadows a variable existing in the outer scope.

practicalswift commented on src/test/addrindex_tests.cpp:
```diff
@@ -0,0 +1,154 @@
+    // Now check that all the addresses we sent to are present in the index.
+    for (int i = 0; i < 10; i++) {
+        std::vector<std::pair<COutPoint, std::pair<CTransactionRef, uint256>>> spends;
+        std::vector<std::pair<COutPoint, std::pair<CTransactionRef, uint256>>> creations;
```
Same here but for `creations`.

marcinja commented:
@Sjors 
> Lightly tested on macOS. It doesn't find the coinbase transaction to `mtnNvZY7iQCfzJCVon13tfJSVYDn9iiUWD` in [block 120](https://testnet.blockchain.info/en/block/0000000058dbb5e3fce9be42ae79beb564be666866c6b85c9791863219ec36db), `mg4bQva8w7Cjs8KaKHeEbyXUaCzoQAcXyH` in [block 10001](https://testnet.blockchain.info/nl/block/0000000000629d100db387f37d0f37c51118f250fb0946310a8c37316cbc4028) and several others.

I'm confused at what you're referring to by`mtn...` and `mg4...`. The outputs for those two coinbase transactions are P2PK, so the only way to search for them is by scriptPubKey.  I used the hex value of the scriptPubKey from [here](https://blockstream.info/block/00000000000080b66c911bd5ba14a74260057311eaeb1982802f7010f1a9f090) and was able to find it.

ryanofsky commented:
Maybe tag this [needing concept ack](https://github.com/bitcoin/bitcoin/labels/Needs%20Conceptual%20Review).

This would probably be pretty easy to rebase, and has a previous concept ack from jimpo, but it's not clear if other people want this change or might object to it.

dan-da commented:
I for one would be very happy to see an optional addr index + api finally available in bitcoin-core, and not as a separate patch or hacky thing.

@marcinja is there an API in your PR that accepts multiple addresses as input and returns total received/spent for each?  Some block explorers have this functionality and it is great because much more efficient than calling API for each addr individually.  

Also I think it should be caller's option (via param) whether to return list of tx or not.  Sometimes only addr summary info is needed eg total received/spent and tx info just adds bloat and slows things down in such cases.

marcinja committed:
Add searchrawtransactions RPC

Adds searchrawtransactions RPC that uses the address index to lookup
transactions and outpoints by script and address. Adds basic functional
tests for searchrawtransactions.

abunsen commented:
@marcinja you're so close. Thanks for all your hard work on this! Does anyone know if this is getting merged anytime soon?

gmaxwell commented:
I haven't looked at how this is implemented, at all-- but the goal is one that is extraordinarily useful and addresses a severe and growing usability probable in the software that has burned me many times.

ryanofsky commented:
> I haven't looked at how this is implemented, at all-- but the goal is one that is extraordinarily useful and addresses a severe and growing usability probable in the software that has burned me many times.

Sorry for the dumb question, but what's the severe usability problem?

ryanofsky commented:
Most recent IRC discussion I could find about address indexes, http://www.erisian.com.au/bitcoin-core-dev/log-2018-12-13.html#l-601 seemed to have conclusion (from @jamesob) with: 
"I'll report on any compelling usecases I find for addr index, but I suspect sipa et al are right that that's usually just the Wrong way"

dan-da commented:
> "usually just the Wrong way"

This sounds like an ivory tower attitude to me.   So tell us wise ones, if an addrindex is the wrong way, then what exactly is the right way?   Because thus far bitcoin-core's answer seems to be:  don't lookup transactions related to an address AT ALL unless they are in your own wallet, which is useless for anyone trying to provide address lookup as a service.  They then must either run a patched bitcoind or turn to btcd or some other 3rd party solution.

ryanofsky commented:
> what exactly is the right way?

The right way to do what? Can you provide more detail? I think it'd be good to keep the discussion as concrete as possible and discuss specific problems and use cases.

dan-da commented:
ok, here are two tools I alone have written that need something like searchrawtransactions().

[hd-wallet-addrs](https://github.com/dan-da/hd-wallet-addrs#blockchain-api-provider-notes) - a tool for discovering all hd-wallet addresses that have ever been used.  So it derives [receive,change] and checks if each address has ever received funds, up to gap limit.  Presently it must rely on 3rd party block explorers APIs.

[bitprices](https://github.com/dan-da/bitprices#blockchain-api-provider-notes) A tool for querying all transactions for all used addresses in a wallet (or any group of addresses) and generating reports including historical prices on the date of each transaction.  Useful for finding gain/loss, etc after the fact.  Presently btcd has the best/fastest implementation, but only after I submitted patches for searchrawtransactions.

Here is a [website frontend](https://mybitprices.info).

Note that the above tools work with any arbitrary seeds or addresses and are expected to return results quickly, so creating a watch-only wallet for each request is not a good option, and anyway it would then just be discarded immediately, so why bother?

The first two links above actually link to discussions of the various API providers and their various strengths/weaknesses for these purposes.

Others have their own reasons.  afaict, most block explorers or bitcoin API providers need some sort of address index and are rolling their own somehow.

ryanofsky commented:
Thanks. From what I can gather the tools described in https://github.com/bitcoin/bitcoin/pull/14053#issuecomment-558344637 are accounting / auditing tools useful for someone running the bitcoin wallet, and this PR would let people who want to use these tools run them locally instead of relying on centralized third party services.

It seems like a good argument in favor of this feature.

I'm not sure what the better arguments against this feature are, so it would be helpful if someone wanted to articulate them. One argument might be that we don't want people getting used to using tools like these, because eventually in the future historical indexes will be less able to scale, and people who are using these tools in their workflow will be forced to turn to centralized third party services.

jnewbery commented:
Concept ACK. The main argument against adding an address index to Bitcoin Core is that people/services will come to rely on them, and they're not scalable long term. My response to that is that people are already using address indexes and block explorers. Providing/not providing that functionality in Bitcoin Core isn't going to change that.

The new indexing infrastructure means that this can be done in a very unobtrusive way. The index has no direct interaction at all with validation or networking. Instead it hooks into the validation interface and runs in a background thread, and so should have no performance impact when the node is sync'ed (I expect this might slow down IBD as `ActivateBestChain()` gets blocked on `LimitValidationInterfaceQueue()`, but presumably users who are indexing the entire transaction history are prepared to wait a bit longer).

jgarzik commented:
addr-index is the most frequently patched-in change to Core, across the years, IME.

dan-da commented:
querying a 3rd party entails a loss of privacy and requires trusting the 3rd party to return truthful data.

jamesob commented:
Concept ACK.

I know I've flip-flopped on this feature historically, but I agree very much with the points that @jnewbery makes. Trading the maintenance burden of ~800 lines of relatively non-consensus-risky indexing code for some very practical gains in user privacy is compelling to me. Even with the caveat that we don't think some of these uses are architecturally sound for years and years.

One might make the argument that there's a slippery slope here - if we consent to this index now, would demand for yet another index come along, and the next and so on? I don't think so. In the few years I've followed this, there has been reliable demand for an address index (and numerous community implementations of varying repute) and nothing else - aside of course from the compact block index - so I don't think a slippery-slope argument holds.

@marcinja do you have time to rebase this PR?

marcinja commented:
Yes, I will rebase this and clean it up a bit. Thanks for the conceptual review.

marcinja committed:
Initialize address index

Setup address index in initialization process.
Add initialization warning and wallet feature request warning as
suggested by ryanofsky.

marcinja committed:
Add address index

Adds index that relates scriptPubKeys to location of transactions in the
filesystem, along with the hash of the block they are found in, and the
outpoint information of the txout with the related script.

marcinja committed:
Add address index test

ryanofsky commented:
Maybe we should log a warning when the address index is enabled, like:

<dl><dd>Warning: built-in address index is enabled! Address indexing is going to become less scalable as transaction history increases, and will eventually need to be removed from [PACKAGE_NAME] and replaced by a dedicated external index. Users relying on the address index for accounting purposes are advised to track metadata in real time so relying on a historical index is not necessary.</dd></dl>

Could also add a coda if the built in wallet is enabled:

<dl><dd>If using the address index to work around lack of tagging or notifications from the built-in [PACKAGE_NAME] wallet, please file wallet feature requests: [PACKAGE_BUGREPORT]</dd></dl>

The idea would be to give end users relying on third party accounting tools a way of knowing about the limitations of those tools, and encouraging authors to implement more scalable approaches.

promag commented:
Concept ACK.

ryanofsky commented:
Nice set of notes on this PR at https://bitcoincore.reviews/14053.html

jnewbery commented on src/rpc/rawtransaction.cpp:
```diff
@@ -213,6 +214,142 @@ static UniValue getrawtransaction(const JSONRPCRequest& request)
+                {
+                    {"address", RPCArg::Type::STR_HEX, RPCArg::Optional::NO, /* default_val */ "", "address or scriptPubKey"},
+                    {"verbose", RPCArg::Type::BOOL, RPCArg::Optional::OMITTED, /* default_val */ "false", "If false, return hex-encoded tx, otherwise return a json object"},
+                    {"count", RPCArg::Type::NUM, RPCArg::Optional::OMITTED, /* default_val */ "", "The block in which to look for the transaction"},
```
"The block in which to look for the transaction" is wrong

andrewtoth replied:
The default value is implicitly 100. It should be explicitly defined here.

jnewbery commented on src/rpc/rawtransaction.cpp:
```diff
@@ -213,6 +214,142 @@ static UniValue getrawtransaction(const JSONRPCRequest& request)
+                {
+                    RPCResult{"if verbose is not set or set to false",
+            "\nResult:\n"
+            " [                                    (array of json objects)\n"
```
I suggest you make this an object with two keys: `creates` and `spends`

jnewbery reviewed (commented):
I've done a preliminary review and left a few comments.

I think this PR could use:

- some design documentation - what's being stored as keys and values in the addr index database and why, and what values are being returned to the RPC user. This could be a code comment at the top of `addrindex.cpp` or `addrindex.h`.
- more test cases, particularly of the spent outputs values (there aren't currently any tests for that) and persistence across stop-start. The functional test can be moved to its own file, rather than being added to `rpc_rawtransaction.py`.

andrewtoth commented on src/rpc/rawtransaction.cpp:
```diff
@@ -213,6 +214,142 @@ static UniValue getrawtransaction(const JSONRPCRequest& request)
 }
 
+static UniValue searchrawtransactions(const JSONRPCRequest& request) {
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 4)
```
```suggestion
    if (request.fHelp || request.params.size() < 1 || request.params.size() > 3)
```

MarcoFalke replied:
Should use the new `.Check()` syntax instead

andrewtoth commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,263 @@
+// to the output spent with the given script, and the transaction it was spent
+// in. creations_result is filled with outpoints for outputs created with this
+// script as their script pubkey, and the transactions they were created in.
+bool AddrIndex::FindTxsByScript(const CScript& script,
```
Shouldn't this method take the count as a parameter and return early when enough entries are found?

andrewtoth commented on src/rpc/rawtransaction.cpp:
```diff
@@ -213,6 +214,142 @@ static UniValue getrawtransaction(const JSONRPCRequest& request)
+        }
+        spends.push_back(tx_val);
+        ++spends_it;
+    }
```
If there are more spends than `count` then no creates will be returned. Ideally the spends and creates will be returned in order, rather than all spends first.

andrewtoth commented on src/rpc/rawtransaction.cpp:
```diff
@@ -213,6 +214,142 @@ static UniValue getrawtransaction(const JSONRPCRequest& request)
     return result;
 }
 
+static UniValue searchrawtransactions(const JSONRPCRequest& request) {
```
There doesn't seem to be a way to skip transactions anymore. Was this removed intentionally?

narula commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,263 @@
+    while (iter->Valid()) {
+        DBKey key;
+        DBValue value;
+        if (!iter->GetKey(key) || key.m_addr_id != addr_id || !iter->GetValue(value) ) break;
```
Do you want to log or throw something here? If I understand the LevelDB docs correctly, you have what LevelDB says is a valid iterator which should be positioned at the key. I think if any of these return false you'd want to know as the DB could be corrupted.

narula replied:
I think in general it would be nice if this class did some of the error logging that I see in `blockfilterindex.cpp`.

narula commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -487,6 +486,44 @@ def run_test(self):
 
+        self._test_searchrawtransactions()
+
+    def _test_searchrawtransactions(self):
```
Please also test the spends_results path (returning transactions which spend from this address).

narula commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,263 @@
+            const CTxUndo& tx_undo = block_undo.vtxundo[i-1];
+            for (size_t k = 0; k < tx.vin.size(); ++k) {
+                CScript spent_outputs_scriptpubkey = tx_undo.vprevout[k].out.scriptPubKey;
+                DBKey key(DBKeyType::SPENT, GetAddrId(spent_outputs_scriptpubkey), tx.vin[k].prevout);
```
I *think* this is correct because he is using the prevout in the *key*. When looking up by this script with this outpoint, one would want to find this transaction because it spends this outpoint.

AFAICT the value just contains the position of the relevant transaction on disk and the scriptPubKey (to detect collisions).  The value does not contain any indexes into inputs or outputs.

narula replied:
Can you add a comment with your key/value format somewhere and the justification?

jonatack reviewed (commented):
Still gathering context here in light of the Concept ACKs others have given. Nevertheless, on first look I'm finding it difficult to ack the concept of investing resources on reviewing, adding, and maintaining a feature that will need to be removed down the road in favor of a dedicated external index by interested parties.

fjahr commented:
For me what this PR is missing some analysis on the current size of the index in mainnet and the growth rate of the data. Has anyone run the numbers recently and can report?

marcinja commented on src/rpc/rawtransaction.cpp:
```diff
@@ -213,6 +214,142 @@ static UniValue getrawtransaction(const JSONRPCRequest& request)
+        UniValue tx_val(UniValue::VOBJ);
+
+        UniValue outpoint_val(UniValue::VOBJ);
+        outpoint_val.pushKV("txid", spend.first.hash.GetHex());
```
```suggestion
        outpoint_val.pushKV("txid", spend.second.first.hash.GetHex());
```

narula commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,263 @@
+        // Place entry into correct vector depending on its type.
+        switch (key.m_key_type) {
+            case DBKeyType::SPENT:
+                spends_result.emplace_back(std::make_pair(key.m_outpoint, result));
```
As pointed out by jnewbery in review club, I think this is wrong. 

The DB's key always represents the outpoint that *created* the scriptPubKey.  For spends_result, you want the txid of the transaction that *spends* the scriptPubKey.  So instead of the DB key's outpoint as the transaction id to be returned to the user, you want the transaction id of the transaction in the DB value.  Or, in the RPC code lines 314 and 315, use the txid from `spend.second`.

fjahr commented:
I think @jnewbery was hinting at this in the PR Review Club as well: it seems like this index does not handle reorgs. `BaseIndex` handles this through `Rewind` which is not overridden.

andrewtoth commented:
Ran on i7-8750H, took about 48 hours on an already synced node. Took 200 GB disk space.
```
2020-01-08T02:22:02Z addcon thread start
2020-01-10T02:37:07Z addr_index is enabled at height 611001
2020-01-10T02:37:07Z addr_index thread exit
```
```
$ du -h addr_index
200G	addr_index
```

luke-jr commented:
Might be a good idea to replace `CDiskTxPos` with a height+tx# for [future?] pruning compatibility, like #13014

luke-jr commented:
> I think @jnewbery was hinting at this in the PR Review Club as well: it seems like this index does not handle reorgs. BaseIndex handles this through Rewind which is not overridden.

Handling reorgs might actually be undesirable for this... but changing CDiskTxPos like I just suggested to a block height could be incompatible with saving references to blocks not in the main chain too. (Maybe storing height OR blockhash would work?)

Talkless commented:
@marcinja Do you believe this index could allow wallets such as Electrum to avoid relying on extra applications (like ElectrumX, Electrum Private Server, etc), and (if leared how to speak Bitcoin RPC) use Bitcoin Core node directly?

baso10 commented:
Any update when this could be merged?

fjahr commented:
> Any update when this could be merged?

I kind of thought that this was still in conceptual review but it has received 5 Concept ACKs from experienced Core developers and I did not see much of the push back this has received historically, so I tend to believe this can be moved out of Conceptual review phase and move forward. There is still a lot of work to be done in terms of testing, documentation etc.

@marcinja If you are currently unable to work on this I would be happy to take this over since I have some experience with indexes already :)

luke-jr commented:
Past NACKs don't disappear just because a new PR is opened...

fjahr commented:
> Past NACKs don't disappear just because a new PR is opened...

Definitely agree, especially if the PRs are very close. But I think if a new PR is opened with a different approach it can still be expected for the reviewers who gave NACKs in the past, to participate in the discussion again and reinstate their NACK or someone else who agrees with their view should do it for them. Otherwise, we might never be able to move forward if a concept is NACKed and then the reviewer leaves the project or just does not comment on the new approach. But it would probably be good if @marcinja would address the past NACKs directly and give those reviewers another chance to say if they have changed their minds or not.

MarcoFalke commented:
Concept ACK

jonatack commented:
> Any update when this could be merged?

@baso10 

1. This PR needs to be rebased, as you can see in https://github.com/bitcoin/bitcoin/pull/14053#issuecomment-610034274

2. The PR is still going through peer review -- see the discussion above

3. See https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review for more info about the peer review process in this project

jonatack commented:
A good resource for more context and history and reasons for and against this addition: https://bitcoincore.reviews/14053.

marcinja commented:
Hey everyone. Just wanted to give a small update here. I am still planning on keeping this PR up-to-date and rebased. I've realized it's more annoying to rebase after procrastinating, and also it's not fair to the reviewers and other people interested so I will be much more diligent about rebasing now.

The total size of the address index right now is 223GB. I don't have a good answer on  index sync time because I was re-syncing on a laptop. This was done on the 8f3fc6b4b8fb1a2db3035398d3dd5a2ed91cc3f9 branch over the course of a few days.  I'm going to time an IBD and addrindex sync on a desktop with a good SSD so as to get meaningful results.

The good news is that `searchrawtransactions` is not prohibitively slow. 
```
time ./src/bitcoin-cli searchrawtransactions 3EiAcrzq1cELXScc98KeCswGWZaPGceT1d true 30000 > big-search

./src/bitcoin-cli searchrawtransactions 3EiAcrzq1cELXScc98KeCswGWZaPGceT1d     4.79s user 1.95s system 49% cpu 13.575 total
```

I have also confirmed that the address index has the same number of created/spent outpoints as on this explorer:  https://blockstream.info/address/3EiAcrzq1cELXScc98KeCswGWZaPGceT1d

fanquake commented:
@marcinja If you feel like it, you could pull 2ce521784432550007c0df1da85b4d3d0b1c7477 and 2761d11cc798d42839676f98250781a3ea445ce2 out of here and PR separately.

romanz commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,279 @@
+ * created as a new  COutpoint in that transaction.
+ *
+ * outpoints are stored in the key as opposed to in the DB value to preserve
+ * uniqueness and to support multiple values for a a single addr_id and key_type
```
```suggestion
 * uniqueness and to support multiple values for a single addr_id and key_type
```

romanz commented on src/index/addrindex.cpp:
```diff
@@ -0,0 +1,279 @@
+ *
+ * The DB keys are structured as follows: <addr_id, key_type, outpoint>
+ *
+ * addr_id is the hash of the script_pub_key computed using MurmurHash3, a a
```
```suggestion
 * addr_id is the hash of the script_pub_key computed using MurmurHash3, a
```

c78867886 commented:
Can someone plz approve this?

decryp2kanon commented:
Concept ACK

Talkless commented:
@marcinja will this allow other wallets like Electrum to utilize this feature (RPC credentials provided, of course) and avoid having to run ElecturmX (https://github.com/spesmilo/electrumx), EPS (https://github.com/chris-belcher/electrum-personal-server) or BWT (https://github.com/shesek/bwt) intermediary software?

@ecdsa @SomberNight @chris-belcher @shesek could you provide your input on how this feature might be useful (or not)?

romanz commented:
The main issue AFAIU is that Electrum is using `SHA256(scriptPubKey)` while this PR is using `MurmurHash3(scriptPubKey)`.
Also, ElectrumX & [electrs](https://github.com/romanz/electrs) are using RocksDB for the index storage - resulting in better performance and disk usage (compared to LevelDB).

SomberNight commented:
@Talkless note that while some Electrum users run their own bitcoind, many do not. Electrum wants to support both use cases, and in fact the suspicion is that most users just use a public server. When using a public server, the client cannot use bitcoind RPC, hence in that case I don't see how the middleware (e.g. ElectrumX) could be avoided.

For the own bitcoind use case, maybe the client could have another optional mode of operation where it uses bitcoind RPC directly, which is I guess what you are asking about. For that, an address-index in bitcoind is the main thing missing indeed, however not the only one. For one, the electrum protocol (the client<->"middleware server" connection) has address subscriptions - the client gets a notification when a history of one its addresses changes. We are also planning on soon adding another method into the protocol that allows `txoutpoint->spender_txid` lookups (and notifications); I guess that could be implemented using the index in this PR albeit in a very inefficient way for heavily reused addresses.

IMHO there are multiple upsides for having this middleware setup for Electrum:
- for the project, it keeps the codebase simpler (again, we want to support users without own bitcoind)
- for the project, it also allows for more flexibility for implementing new functionality: just consider present PR here, we have needed such an index for 9 years but bitcoind did not have it or want it, so we could just implement it ourselves
- for the server operator, even if they don't want to open up the server for the public, they could share it with their friends and family. I don't think that's feasible with bitcoind RPC. I think this is a common use case.

Nevertheless if someone steps up and contributes patches, this kind of thing could be added.

-----

> The total size of the address index right now is 223GB

That sounds much larger than expected.
Even with the `txoutpoint->spender_txid` map I mentioned above, when using LevelDB, ElectrumX uses around 90 GiB of disk space.

> The DB keys are structured as follows: `<addr_id, key_type, outpoint>`
> The DB values are simply: `<CDiskTxPos, CScript> `

Instead of having both `addr_id` and `CScript`, why not just put a long hash, e.g. `sha256(CScript)` into the key?

Also, I expect most users of this index would also want txindex enabled. You might want to consider making address index dependent on txindex. Have you investigated how much space that would save? There would be no need to store `CDiskTxPos`.

Another trick that ElectrumX uses is that only the best chain is indexed. We have a `tx_num->txid` map as raw files on disk. A `tx_num` is the index of the transaction in the linear history of the blockchain, a 5-byte integer (so e.g. the genesis block coinbase tx has `tx_num=0`). This map uses around 17 GiB atm. With this, you can encode the txid part of the `outpoint` as 5 bytes instead of 32 bytes.

MarcoFalke commented:
Concept ACK (might have already done that)

sipa commented:
I'm concept -0 on this.

My primary objection is that I think it's a bad idea for any infrastructure to be built that relies on having fully-indexed blockchain data available.

However, it seems many people want something like this, and are going to use it anyway. The question is then whether it belongs in the bitcoin-core codebase. Alternative, and more performant presumably, like electrs exist already too, so it isn't exactly impossible to do this elsewhere.

Still, given that we now have the indexes infrastructure, it means that things like this are easy to add in a fairly modular way without invading consensus code. So if people really want this, fine.

jamesob commented:
I'm also a little more negative on having this in Core than I previously was. After working in a few industrial contexts on wallet stuff, it's clear to me that an address index is really only required if you want to implement a block explorer or do chain analysis. For both of these applications, using something like electrs seems sufficient.

For personal wallet management, a full address index is not required. I think the origin of some confusion is that things like the Electrum Personal Server have become synonymous with this kind of usage, but in reality a full index is overkill when descriptor-specific rescans can be done once for a historical backfill and then per-block scanning can be done from there on out.

I want to point out that this is a nice implementation and good work by @marcinja, but I'm leaning slightly against the inclusion of such an index in Core at this point.

jonasschnelli commented:
> My primary objection is that I think it's a bad idea for any infrastructure to be built that relies on having fully-indexed blockchain data available.

I agree on this.

IMO the only use cases to ever use a full address index are:
* 1) Instant seed/xpriv backup recovery including spent history
* 2) Backend service for thousands of wallets
* 3) Debug/explore purposes

**1** (instant backup recovery) could be solved with either `scantxoutset` (take a minute or two) or by an address-index for the utxo set only. But both would not restore the spent history.
A scalable non-enterprise solution to restore the spent history is using blockfilters. Scan through the filters and rescan only the relevant blocks (a matter of minutes), see #20664.

**2** (a backend for thousands of wallets): out of scope for this project.

**3** (explore purposes): I think this is a valid use case. Though adding this PR to Bitcoin Core will lead to **many many projects** using it **in production** increasing the traffic in this project and eventually *steal* time from existing contributors (rebase, maintenance, drag-along)

**My main fear is that people are going to use this index (a full address index) to use it as an electrum(ish) backend for a handful of wallets.**

With multiwallet, watch-only-wallets, PSBT, we have all tools to server multiple wallets in a scalable way for external applications.

I also think merging this as it is, would be in contradiction to the process- and repository-separation effort.

Therefore I'm ~0 (slightly towards NACK) to add this.
If this would be in another repository (still under `bitcoin/*`) and process separated, I would ACK it.

marcinja commented:
Hi all, thanks for the feedback and review. This was an enjoyable PR to work on and I learned a lot from all your comments.

I'm closing this PR because its size probably requires stronger support from contributors to get in. It also seems more clear now that all of the practical use-cases are covered by existing features and some lightweight alternatives (https://github.com/bitcoin/bitcoin/pull/20664) .

I also agree that it would be bad to incentivize using an address index to support an external electrum wallet, when it's not the intended use-case and would cause unnecessary burden on contributors and maintainers in this project, e.g. from users of those wallets wanting new features or updates.

jonatack commented:
@marcinja thank you and I hope to see more contributions of this quality from you.
